tests/unit/test_comparator.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Inspirehep.
+# Copyright (C) 2016 CERN.
+#
+# Inspirehep is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Inspirehep is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Inspirehep; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+
+"""Test merger corner cases that are unlikely to appear as usage scenarios."""
+
+from __future__ import absolute_import, print_function
+
+from json_merger.comparator import PrimaryKeyComparator
+
+
+def test_multiple_primary_keys():
+    class MyComp(PrimaryKeyComparator):
+        primary_key_fields = ['id', 'f.id']
+
+    lst = [{'id': 0}, {'id': 1}, {'f': {'id': 0}}, {'f': {'id': 1}}]
+    inst = MyComp(lst, lst)
+
+    for idx1 in range(len(lst)):
+        for idx2 in range(len(lst)):
+            if idx1 == idx2:
+                assert inst.equal(idx1, idx2)
+            else:
+                assert not inst.equal(idx1, idx2)
+
+
+def test_list_of_primary_keys():
+    class MyComp(PrimaryKeyComparator):
+        primary_key_fields = [['id1', 'id2']]
+
+    lst1 = [{'id1': 0, 'data': 1},
+            {'id2': 0, 'data': 1},
+            {'id1': 1, 'id2': 1, 'data': 1}]
+    lst2 = [{'id1': 0, 'data': 2},
+            {'id2': 0, 'data': 2},
+            {'id1': 1, 'id2': 1, 'data': 2},
+            {'id1': 1, 'id2': 0, 'data': 2}]
+
+    inst = MyComp(lst1, lst2)
+
+    assert not inst.equal(0, 0)
+    assert not inst.equal(1, 1)
+    assert not inst.equal(2, 3)
+
+    assert inst.equal(2, 2)
```

Comment:
The `PrimaryKeyComparator` needs to be sublcassed so we can compare by different fields. The class will be instantiated in the merger so there's no need to actually use it.

An use-case for it can be found here:
https://github.com/mihaibivol/json-merger-viewer/blob/master/merger_config.py

The reason why comparator classes work with indices is because we might want to memoize some comparisons (for example in author lists where the matching function is pretty heavy) and it's easier to memoize by index than by dict.